library/core/src/iter/adapters/mod.rs

```diff
@@ -1687,6 +1687,110 @@ impl<I: Iterator> Peekable<I> {
     {
         self.next_if(|next| next == expected)
     }
+
+    /// Creates an iterator that consumes elements until predicate is
+    /// true, without consuming the last matching element.
+    ///
+    /// `until()` takes a closure as an argument. It will call this
+    /// closure on each element of the iterator, and consume elements
+    /// until it returns true.
+    ///
+    /// After true is returned, until()'s job is over, and the iterator
+    /// is fused.
+    ///
+    /// This is the exact opposite of [`skip_while`].
+    ///
+    /// # Example
+    /// Consume numbers until you find a '5'.
+    /// ```
+    /// #![feature(peekable_next_if)]
+    /// let mut iter = (0..10).peekable();
+    /// assert_eq!(iter.until(|&x| x == 5).collect::<String>(), "1234".to_string());
+    /// assert_eq!(iter.next(), Some(5));
+    /// ```
+    /// [`skip_while`]: trait.Iterator.html#method.skip_while
+    #[unstable(feature = "peekable_next_if", issue = "72480")]
+    pub fn until<P: FnMut(&I::Item) -> bool>(&mut self, func: P) -> Until<'_, I, P> {
+        Until::new(self, func)
+    }
+}
+
+/// An iterator that iterates over elements until `predicate` returns `false`.
+///
+/// This `struct` is created by the [`until`] method on [`Peekable`]. See its
+/// documentation for more.
+///
+/// [`until`]: trait.Peekable.html#until
+/// [`Iterator`]: trait.Iterator.html
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "peekable_next_if", issue = "72480")]
+pub struct Until<'a, I, P>
+where
+    I: Iterator,
+{
+    peekable: &'a mut Peekable<I>,
+    flag: bool,
+    predicate: P,
+}
+impl<'a, I, P> Until<'a, I, P>
+where
+    I: Iterator,
+{
+    fn new(peekable: &'a mut Peekable<I>, predicate: P) -> Self {
+        Until { peekable, flag: false, predicate }
+    }
+}
+
+#[stable(feature = "core_impl_debug", since = "1.9.0")]
+impl<I, P> fmt::Debug for Until<'_, I, P>
+where
+    I: fmt::Debug + Iterator,
+    I::Item: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Until").field("peekable", self.peekable).field("flag", &self.flag).finish()
+    }
+}
+
+#[unstable(feature = "peekable_next_if", issue = "72480")]
+impl<I: Iterator, P> Iterator for Until<'_, I, P>
+where
+    P: FnMut(&I::Item) -> bool,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<I::Item> {
```

Comment:
`skip_while` uses `fn check` most likely to optimized the assembly, how does this assembly looks like? @lzutao do you think this will have an impact? `SkipWhile`
```rust
impl<I: Iterator, P> Iterator for SkipWhile<I, P>
where
    P: FnMut(&I::Item) -> bool,
{
    type Item = I::Item;

    #[inline]
    fn next(&mut self) -> Option<I::Item> {
        fn check<'a, T>(
            flag: &'a mut bool,
            pred: &'a mut impl FnMut(&T) -> bool,
        ) -> impl FnMut(&T) -> bool + 'a {
            move |x| {
                if *flag || !pred(x) {
                    *flag = true;
                    true
                } else {
                    false
                }
            }
        }

        let flag = &mut self.flag;
        let pred = &mut self.predicate;
        self.iter.find(check(flag, pred))
    }
```